libbpf-tools/tcpconnect.bpf.c

```diff
@@ -0,0 +1,217 @@
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 uid = bpf_get_current_uid_gid();
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = pid_tgid;
+
+	if (filter_pid && pid != filter_pid)
+		return 0;
+
+	uid = bpf_get_current_uid_gid();
+	if (filter_uid != (uid_t) -1 && uid != filter_uid)
+		return 0;
+
+	bpf_map_update_elem(&sockets, &tid, &sk, 0);
+	return 0;
+}
+
+static  __always_inline void count_v4(struct sock *sk, __u16 dport)
+{
+	struct ipv4_flow_key key = {};
+	static __u64 zero;
+	volatile __u64 *val;
+
+	BPF_CORE_READ_INTO(&key.saddr, sk, __sk_common.skc_rcv_saddr);
+	BPF_CORE_READ_INTO(&key.daddr, sk, __sk_common.skc_daddr);
+	key.dport = dport;
+	val = bpf_map_lookup_or_try_init(&ipv4_count, &key, &zero);
+	if (val)
+		*val += 1;
+}
+
+static __always_inline void count_v6(struct sock *sk, __u16 dport)
+{
+	struct ipv6_flow_key key = {};
+	static const __u64 zero;
+	volatile __u64 *val;
+
+	BPF_CORE_READ_INTO(&key.saddr, sk,
+			   __sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+	BPF_CORE_READ_INTO(&key.daddr, sk,
+			   __sk_common.skc_v6_daddr.in6_u.u6_addr32);
+	key.dport = dport;
+
+	val = bpf_map_lookup_or_try_init(&ipv6_count, &key, &zero);
+	if (val)
+		*val += 1;
+}
+
+static __always_inline void
+trace_v4(struct pt_regs *ctx, pid_t pid, struct sock *sk, __u16 dport)
+{
+	struct event event = { .af = AF_INET, };
+
+	event.pid = pid;
+	event.uid = bpf_get_current_uid_gid();
+	event.ts_us = bpf_ktime_get_ns() / 1000;
+	BPF_CORE_READ_INTO(&event.saddr_v4, sk, __sk_common.skc_rcv_saddr);
+	BPF_CORE_READ_INTO(&event.daddr_v4, sk, __sk_common.skc_daddr);
+	event.dport = dport;
+	bpf_get_current_comm(event.task, sizeof(event.task));
+
+	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU,
+			      &event, sizeof(event));
+}
+
+static __always_inline void
+trace_v6(struct pt_regs *ctx, pid_t pid, struct sock *sk, __u16 dport)
+{
+	struct event event = { .af = AF_INET6, };
+
+	event.pid = pid;
+	event.uid = bpf_get_current_uid_gid();
+	event.ts_us = bpf_ktime_get_ns() / 1000;
+	BPF_CORE_READ_INTO(&event.saddr_v6, sk,
+			   __sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+	BPF_CORE_READ_INTO(&event.daddr_v6, sk,
+			   __sk_common.skc_v6_daddr.in6_u.u6_addr32);
+	event.dport = dport;
+	bpf_get_current_comm(event.task, sizeof(event.task));
+
+	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU,
+			      &event, sizeof(event));
+}
+
+static __always_inline int
+exit_tcp_connect(struct pt_regs *ctx, int ret, int ip_ver)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u32 tid = pid_tgid;
+	struct sock **skpp;
+	struct sock *sk;
+	__u16 dport;
+
+	skpp = bpf_map_lookup_elem(&sockets, &tid);
+	if (!skpp)
+		return 0;
+
+	if (ret)
+		goto end;
+
+	sk = *skpp;
+
+	BPF_CORE_READ_INTO(&dport, sk, __sk_common.skc_dport);
+	dport = bpf_ntohs(dport);
+	if (filter_port(dport))
+		goto end;
+
+	if (do_count)
+		((ip_ver == 4) ? count_v4 : count_v6)(sk, dport);
```

Comment:
ugh, IMO, too clever, let's keep it simple with if/else? and unnecessary () around `ip_ver == 4`

Reply:
done